src/main/java/com/rackspacecloud/metrics/ingestionservice/influxdb/InfluxDBHelper.java

```diff
@@ -269,18 +267,16 @@ private boolean createDatabase(final String databaseName, final String baseUrl,
     private InfluxDB getInfluxDBClient(String instanceUrl) {
         InfluxDB influxDB = this.urlInfluxDBInstanceMap.get(instanceUrl);
 
+        if (influxDB != null) return influxDB;
+
         BatchOptions batchOptions = BatchOptions.DEFAULTS;
         batchOptions.actions(this.numberOfPointsInAWriteBatch);
         batchOptions.flushDuration(this.writeFlushDurationMsLimit);
-//        batchOptions.bufferLimit(2000);
-
-        if(influxDB == null) {
-            influxDB = this.influxDBUtils.getInfluxDB(instanceUrl);
-            influxDB.setLogLevel(InfluxDB.LogLevel.BASIC);
-//            influxDB.enableBatch(BatchOptions.DEFAULTS);
-            influxDB.enableBatch(batchOptions);
-            this.urlInfluxDBInstanceMap.put(instanceUrl, influxDB);
-        }
+
+        influxDB = this.influxDBFactory.getInfluxDB(instanceUrl);
+        influxDB.setLogLevel(InfluxDB.LogLevel.BASIC);
+        influxDB.enableBatch(batchOptions);
+        this.urlInfluxDBInstanceMap.put(instanceUrl, influxDB);
```

Comment:
Using the `get` and `put` as separate calls loses the concurrency guarantees of `ConcurrentHashMap`. Instead use this

https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentMap.html#computeIfAbsent-K-java.util.function.Function-

which will atomically guarantee to either get the existing value or put the computed value if not present.

The race condition is discussed more here https://stackoverflow.com/a/14947844/121324

Reply:
Done